### PR TITLE
Adding Team functionality in Asana 

### PR DESCRIPTION
**Pull request by stanty522**
## Description

Made changes to the Asana extension to include Teams. This creates a filter for Projects to only be searched under the selected Team.

This addresses an earlier issue where the query for Projects would time out if a workspace had too many existing Projects.

## Screencast

CleanShot 2022-10-17 at 14 31 14

## Checklist

[ ✅] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
[ ✅] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
[ ✅] I ran npm run build and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
[✅ ] I checked that files in the assets folder are used by the extension itself
[ ✅] I checked that assets used by the README are placed outside of the metadata folder

### CODE REVIEW

File: extensions/asana/src/api/teams.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { request } from "./request";
+
+export type Teams = {
+  gid: string;
+  name: string;
+};
+
+export async function getTeamsForWorkspace(workspace: 1201518352966289) {
```
Comment (thomaslombart) [inline extensions/asana/src/api/teams.ts:8]: I'm not sure the hardcoded number is supposed to be there.

File: extensions/asana/src/components/CreateTaskForm.tsx [hunk 2]
```diff
@@ -126,12 +132,24 @@ export default function CreateTaskForm(props: {
...
       enableDrafts={!props.fromEmptyView}
     >
-      <Form.Dropdown title="Workspace" storeValue {...itemProps.workspace}>
+      <Form.Dropdown title="Workspace" storeValue  {...itemProps.workspace}>
         {workspaces?.map((workspace) => {
           return <Form.Dropdown.Item key={workspace.gid} value={workspace.gid} title={workspace.name} />;
         })}
       </Form.Dropdown>
 
+      <Form.Dropdown title="Teams" storeValue autoFocus {...itemProps.team}>
```
Comment (thomaslombart) [inline extensions/asana/src/components/CreateTaskForm.tsx:141]: If `storeValue` is used for teams, then the `autofocus` should be kept for the "Task Name" field as the team value will be remembered for the subsequent creations.

**Review by thomaslombart — COMMENTED** [review]
I've tried it on my test workspace and it seemed to work but I'm a bit confused about how the teams are supposed to work when you don't have an organization in Asana.

Here's what I have in Asana:

CleanShot 2022-10-18 at 10 44 04@2x

As you can see, I don't have teams. So, I'm wondering if the teams field should not be there only if the users do have teams in their workspace.

File: extensions/asana/src/api/teams.ts (hunk 1 above)
Comment (stanty522) [reply extensions/asana/src/api/teams.ts:8]: oop! yeah my bad there.

File: extensions/asana/src/components/CreateTaskForm.tsx (hunk 2 above)
Comment (stanty522) [reply extensions/asana/src/components/CreateTaskForm.tsx:141]: I tried putting autoFocus on the Task Name like you said but it still wouldn't remember the value for Teams or Projects for the next time I create a task.

Comment (thomaslombart) [reply extensions/asana/src/components/CreateTaskForm.tsx:141]: And it does remember the value when there's `autoFocus` on the projects field? It doesn't seem related to me 🤔

**Comment by thomaslombart** [comment]
Oh, didn't know that. Then I guess, the team's field can simply be hidden if there's only one team available for the user and if the experience is the same when you add a task to a team instead of a workspace for a personal use 🙂

**Comment by stanty522** [comment]
> 

Hey here's my attempt at trying to fix this. Unfortunately I'm running into some basic syntax error :(

However I think the logic will work. Basically in the Asana documentation there is a boolean property for workspaces that informs us if its an organization or not. So basically I've added this to the workspaces.ts and then I plan to use this as a variable in the CreatTaskForm.tsx which will be a conditional if statement to return the form drop down for teams or not. If true (i,e, workspace is an organization) then return teams form drop down, else don't return. 


  ```
const isOrganization = workspaces?.map((workspace) => workspace.is_organization);

    if (isOrganization === true) {
        return 
          <Form.Dropdown title="Teams" storeValue {...itemProps.team}>
          {teams?.map((teams) => {
            return (
            <Form.Dropdown.Item 
              key={teams.gid} 
              value={teams.gid} 
              title={teams.name} 
              />
            );
          })}
        </Form.Dropdown>
```

**Comment by stanty522** [comment]
@thomaslombart 

I figured out the logic of checking if the workspace is an organisation or not. While this works, it unfortunately creates a new issue whereby the workspaces that are NOT organisations cannot select a team, and if a team is not selected a project cannot be selected.

`        {workspaces?.map((workspace) => {
            if (workspace.is_organization) {
              return (
                <Form.Dropdown title="Teams" storeValue {...itemProps.team}>
                {teams?.map((teams) => {
                  return (
                  <Form.Dropdown.Item 
                    key={teams.gid} 
                    value={teams.gid} 
                    title={teams.name} 
                    />
                  );
                })}
              </Form.Dropdown>
              );
            }
          })
        }

      <Form.TagPicker title="Projects" placeholder="Select one or more projects" storeValue {...itemProps.projects}>
        {allProjects?.map((project) => {
          return (
            <Form.TagPicker.Item
              key={project.gid}
              icon={getProjectIcon(project)}
              title={project.name}
              value={project.gid}
            />
          );
        })}
      </Form.TagPicker>`

**Comment by thomaslombart** [comment]
> and if a team is not selected a project cannot be selected.

But in that case, the projects can be retrieved based on a workspace instead of a team?

**Comment by stanty522** [comment]
> But in that case, the projects can be retrieved based on a workspace instead of a team?

Yes that's possible but I'm not sure how I'd go about that without over complicating the structure :(

**Comment by thomaslombart** [comment]
Then, I'll take a look at that probably end of next week. I'm focusing on another project at the moment 😊

**Comment by thomaslombart** [comment]
After taking a look, I think it can be done in a few steps:
- `useProjects` and `getProjects` should be modified to handle an additional `workspace` parameter. The algorithm would be pretty simple: the team is used first to retrieve the projects. But, if it's undefined (in the case you mentioned), then the workspace is used to retrieve it.
- Then in `CreateTaskForm`, you could pass the workspace and the team to `useProjects`: `useProjects({ workspace, team })`

Otherwise, someone from the Raycast team told me the API call to retrieve the projects in the native extension was different than the one used here. Here we use the `/projects` endpoint whereas in the native extension it's `/workspaces/{workspaceId}/typeahead` with the `resource_type` param specified as `project`. It may be easier to switch to it to remove the project's pagination issue here.

**Comment by stanty522** [comment]
So I tried the typeahead thing first, and I'm not sure if I did it correctly, but I just ended up with the same issue of having too many projects and for it to even work I had to put the optional parameter of `limit`. I'm not sure if there's a way to only start the query after X characters have been typed so it won't try to query everything from the start. 

The steps you suggested, I only got as far as to add the additional `workspace` parameter to `getProjects` and `useProjects`. Im still quite new to typescript and async programming so not exactly sure how to implement the if then statement check that would retrieve workspace. Is there a similar piece of code where this is done so I can learn from that?

**Comment by thomaslombart** [comment]
Hey there, I'm not sure where this could have been done... 
Unfortunately, I won't have enough bandwidth to work on that this week. I'll take another look at it probably next week or maybe the week after.

**Pull request closed without merging**